state/presence/pingbatcher.go

```diff
@@ -0,0 +1,266 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+package presence
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/juju/errors"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	"gopkg.in/tomb.v1"
+)
+
+const maxBatch = 1000
+
+type slot struct {
+	Slot  int64
+	Alive map[string]uint64
+}
+
+type singlePing struct {
+	Slot      int64
+	ModelUUID string
+	FieldKey  string
+	FieldBit  uint64
+}
+
+// NewPingBatcher creates a worker that will batch ping requests and prepare them
+// for insertion into the Pings collection. Pass in the base "presence" collection.
+// flushInterval is how often we will write the contents to the database.
+// It should be shorter than the 30s slot window for us to not cause active
+// pingers to show up as missing. The current default is 1s as it provides a good
+// balance of significant-batching-for-performance while still having responsiveness
+// to agents coming alive.
+// Note that we don't strictly sync on flushInterval times, but use a range of
+// times around that interval to avoid having all ping batchers get synchronized
+// and still be issuing all requests concurrently.
+func NewPingBatcher(base *mgo.Collection, flushInterval time.Duration) *PingBatcher {
+	var pings *mgo.Collection
+	if base != nil {
+		pings = pingsC(base)
+	}
+	pb := &PingBatcher{
+		pings:         pings,
+		pending:       make(map[string]slot),
+		flushInterval: flushInterval,
+		pingChan:      make(chan singlePing),
+		flushChan:     make(chan chan struct{}),
+	}
+	pb.start()
+	return pb
+}
+
+// NewDeadPingBatcher returns a PingBatcher that is already stopped with an error.
+func NewDeadPingBatcher(err error) *PingBatcher {
+	// we never start the loop, so the timeout doesn't matter.
+	pb := &PingBatcher{
+		pings:         nil,
+		pending:       make(map[string]slot),
+		flushInterval: time.Millisecond,
+		flushChan:     make(chan chan struct{}),
+	}
+	pb.tomb.Kill(err)
+	pb.tomb.Done()
+	return pb
+}
+
+// PingBatcher aggregates several pingers to update the database on a fixed schedule.
+type PingBatcher struct {
+	pings         *mgo.Collection
+	modelUUID     string
+	pending       map[string]slot
+	pingCount     uint64
+	flushInterval time.Duration
+	tomb          tomb.Tomb
+	pingChan      chan singlePing
+	flushChan     chan chan struct{}
+}
+
+// Start the worker loop.
+func (pb *PingBatcher) start() {
+	rand.Seed(time.Now().UnixNano())
```

Comment:
I'm pretty sure this is silly. You only want to seed at import time, not every time a PingBatcher is created.

For cleanliness, I'd have a private `rand.Rand` instance for use by the package which is just seeded once at import time.